File: crates/quip-protocol/src/session.rs
//! Miner session handshake: `Hello`/`Welcome`/`Configure` and exit codes.

use quip_proto::v1::{Algorithm, Backend, Capabilities, Configure, Hello, Welcome};

/// Protocol version this SDK speaks. `Welcome.protocol_version` must equal this.
pub const PROTOCOL_VERSION: u32 = 2;

/// Errors from building a `Hello` or validating a `Welcome`.
///
/// `Copy`, because the session loop both reports one of these and derives the
/// process exit code from it ([`ExitCode::from`]), and moving it into the
/// conversion would leave nothing to report.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SessionError {
    /// `QUIP_SESSION_TOKEN` missing or empty.
    MissingToken,
    /// `Welcome.protocol_version` did not match [`PROTOCOL_VERSION`].
    BadWelcome(u32),
}

impl std::fmt::Display for SessionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::MissingToken => {
                write!(f, "QUIP_SESSION_TOKEN environment variable not set")
            }
            Self::BadWelcome(v) => write!(f, "unexpected protocol version in Welcome: {v}"),
        }
    }
}

impl std::error::Error for SessionError {}

/// Sysexits-style process exit codes (also carried on `Fatal.exit_code`).
#[repr(i32)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ExitCode {
    /// Clean exit.
    Clean = 0,
    /// Missing/invalid CLI or config (e.g. no `--quip-coordinator`, bad Welcome).
    ConfigInvalid = 64,
    /// Host/env cannot run this miner (`--check` failed).
    EnvIncompatible = 69,
    /// Unexpected internal failure.
    InternalFatal = 70,
    /// `QUIP_SESSION_TOKEN` missing/empty or rejected.
    TokenRejected = 77,
}

impl ExitCode {
    /// Integer value of this exit code (for process exit / `Fatal.exit_code`).
    #[must_use]
    pub fn as_i32(self) -> i32 {
        self as i32
    }
}

impl From<SessionError> for ExitCode {
    fn from(e: SessionError) -> Self {
        match e {
            SessionError::MissingToken => Self::TokenRejected,
            SessionError::BadWelcome(_) => Self::ConfigInvalid,
        }
    }
}

/// Runtime session parameters after applying `Configure` defaults.
pub struct SessionConfig {
    /// Miner identity string from the local config.
    pub miner_id: String,
    /// Max in-flight jobs; default `3` when configure sends `0`.
    pub queue_depth: u32,
    /// Idle timeout seconds; default `300` when configure sends `0`.
    pub idle_timeout_s: u32,
    /// Heartbeat interval seconds; default `15` when configure sends `0`.
    pub heartbeat_s: u32,
    /// Reconnect window seconds; default `60` when configure sends `0`.
    pub reconnect_window_s: u32,
}

impl SessionConfig {
    /// Build config from a miner id and coordinator `Configure`, applying
    /// defaults for any zero field.
    #[must_use]
    pub fn from_configure(miner_id: String, c: &Configure) -> Self {
        let d = |v: u32, default: u32| if v == 0 { default } else { v };
        Self {
            miner_id,
            queue_depth: d(c.queue_depth, 3),
            idle_timeout_s: d(c.idle_timeout_s, 300),
            heartbeat_s: d(c.heartbeat_s, 15),
            reconnect_window_s: d(c.reconnect_window_s, 60),
        }
    }
}

/// Build the miner handshake, reading the session token from the environment.
///
/// # Errors
/// Returns [`SessionError::MissingToken`] if the token is missing or empty.
pub fn build_hello(miner_id: &str, capabilities: Capabilities) -> Result<Hello, SessionError> {
    let token = std::env::var("QUIP_SESSION_TOKEN").map_err(|_| SessionError::MissingToken)?;
    if token.is_empty() {
        return Err(SessionError::MissingToken);
    }
    Ok(Hello {
        miner_id: miner_id.into(),
        session_token: token,
        capabilities: Some(capabilities),
    })
}

/// Reject a `Welcome` whose `protocol_version` is not what this SDK speaks.
///
/// # Errors
/// Returns [`SessionError::BadWelcome`] when the version is not
/// [`PROTOCOL_VERSION`].
pub fn check_welcome(w: &Welcome) -> Result<(), SessionError> {
    if w.protocol_version != PROTOCOL_VERSION {
        return Err(SessionError::BadWelcome(w.protocol_version));
    }
    Ok(())
}

const BACKEND_NAMES: &[(Backend, &str)] = &[
    (Backend::Unspecified, "unspecified"),
    (Backend::Cpu, "cpu"),
    (Backend::Cuda, "cuda"),
    (Backend::Metal, "metal"),
    (Backend::Ane, "ane"),
    (Backend::DwaveQpu, "dwave-qpu"),
    (Backend::Exec, "exec"),
    (Backend::Mock, "mock"),
];
/// Stable lowercase name for a wire identity.
#[must_use]
pub fn backend_name(value: Backend) -> &'static str {
    BACKEND_NAMES
        .iter()
        .find_map(|&(v, name)| (v == value).then_some(name))
        .unwrap_or("unspecified")
}
/// Parse a stable lowercase wire identity name.
#[must_use]
pub fn backend_from_name(name: &str) -> Option<Backend> {
    BACKEND_NAMES
        .iter()
        .find_map(|&(v, n)| (n == name).then_some(v))
}

const ALGORITHM_NAMES: &[(Algorithm, &str)] = &[
    (Algorithm::Unspecified, "unspecified"),
    (Algorithm::Sa, "sa"),
    (Algorithm::Gibbs, "gibbs"),
    (Algorithm::QuantumAnneal, "quantum-anneal"),
    (Algorithm::Fsa, "fsa"),
    (Algorithm::Msa, "msa"),
    (Algorithm::Flatiron, "flatiron"),
    (Algorithm::Mps, "mps"),
    (Algorithm::Mfa, "mfa"),
    (Algorithm::Sb, "sb"),
    (Algorithm::Bsb, "bsb"),
    (Algorithm::Gbsb, "gbsb"),
    (Algorithm::Gdsb, "gdsb"),
    (Algorithm::Ggdsb, "ggdsb"),
    (Algorithm::Hbsb, "hbsb"),
    (Algorithm::Hdsb, "hdsb"),
    (Algorithm::Sbqa, "sbqa"),
    (Algorithm::Tedsb, "tedsb"),
    (Algorithm::External, "external"),
];
/// Stable lowercase name for a wire identity.
#[must_use]
pub fn algorithm_name(value: Algorithm) -> &'static str {
    ALGORITHM_NAMES
        .iter()
        .find_map(|&(v, name)| (v == value).then_some(name))
        .unwrap_or("unspecified")
}
/// Parse a stable lowercase wire identity name.
#[must_use]
pub fn algorithm_from_name(name: &str) -> Option<Algorithm> {
    ALGORITHM_NAMES
        .iter()
        .find_map(|&(v, n)| (n == name).then_some(v))
}

#[cfg(test)]
mod tests {
    use super::*;
    use quip_proto::v1::{Algorithm, Backend, Capabilities, Configure, Welcome};

    #[test]
    fn hello_requires_token() {
        let caps = Capabilities {
            protocol_version: 2,
            features: vec!["streaming".into()],
            ..Default::default()
        };
        std::env::remove_var("QUIP_SESSION_TOKEN");
        assert_eq!(
            build_hello("cpu-0", caps.clone()),
            Err(SessionError::MissingToken)
        );
        std::env::set_var("QUIP_SESSION_TOKEN", "tok-123");
        let hello = build_hello("cpu-0", caps.clone()).unwrap();
        assert_eq!(hello.session_token, "tok-123");
        assert_eq!(hello.capabilities, Some(caps.clone()));
        assert_eq!(PROTOCOL_VERSION, 2);
        std::env::set_var("QUIP_SESSION_TOKEN", "");
        assert_eq!(build_hello("cpu-0", caps), Err(SessionError::MissingToken));
        std::env::remove_var("QUIP_SESSION_TOKEN");
    }

    #[test]
    fn identity_names_are_unique_and_roundtrip() {
        let mut names = std::collections::HashSet::new();
        for raw in 0..=7 {
            let value = Backend::try_from(raw).unwrap();
            assert!(names.insert(backend_name(value)));
            assert_eq!(backend_from_name(backend_name(value)), Some(value));
        }
        names.clear();
        for raw in 0..=18 {
            let value = Algorithm::try_from(raw).unwrap();
            assert!(names.insert(algorithm_name(value)));
            assert_eq!(algorithm_from_name(algorithm_name(value)), Some(value));
        }
        assert_eq!(backend_from_name("test"), None);
        assert_eq!(algorithm_from_name("quantum"), None);
    }

    #[test]
    fn configure_applies_defaults_for_zero_fields() {
        let c = Configure {
            queue_depth: 0,
            idle_timeout_s: 0,
            heartbeat_s: 0,
            reconnect_window_s: 0,
            backend_toml: String::new(),
        };
        let cfg = SessionConfig::from_configure("cpu-0".into(), &c);
        assert_eq!(cfg.queue_depth, 3);
        assert_eq!(cfg.idle_timeout_s, 300);
        assert_eq!(cfg.heartbeat_s, 15);
        assert_eq!(cfg.reconnect_window_s, 60);
    }

    #[test]
    fn welcome_rejects_non_v2_protocol_version() {
        assert!(check_welcome(&Welcome {
            protocol_version: 2
        })
        .is_ok());
        assert_eq!(
            check_welcome(&Welcome {
                protocol_version: 1
            }),
            Err(SessionError::BadWelcome(1))
        );
        assert_eq!(
            check_welcome(&Welcome {
                protocol_version: 0
            }),
            Err(SessionError::BadWelcome(0))
        );
    }

    #[test]
    fn session_error_maps_to_documented_exit_codes() {
        assert_eq!(
            ExitCode::from(SessionError::MissingToken),
            ExitCode::TokenRejected
        );
        assert_eq!(
            ExitCode::from(SessionError::BadWelcome(2)),
            ExitCode::ConfigInvalid
        );
        assert_eq!(ExitCode::ConfigInvalid.as_i32(), 64);
        assert_eq!(ExitCode::EnvIncompatible.as_i32(), 69);
        assert_eq!(ExitCode::InternalFatal.as_i32(), 70);
        assert_eq!(ExitCode::TokenRejected.as_i32(), 77);
    }
}

File: python/quip_solver_core/session.py
import os
from dataclasses import dataclass

from quip.v1 import miner_pb2

# Protocol version this SDK speaks. Welcome.protocol_version must equal this.
PROTOCOL_VERSION = 2

# Stable lowercase names for the wire identity enums. The strings are the
# quip_protocol::session maps (backend_name / algorithm_name).
BACKEND_NAMES: dict[int, str] = {
    miner_pb2.BACKEND_UNSPECIFIED: "unspecified",
    miner_pb2.BACKEND_CPU: "cpu",
    miner_pb2.BACKEND_CUDA: "cuda",
    miner_pb2.BACKEND_METAL: "metal",
    miner_pb2.BACKEND_ANE: "ane",
    miner_pb2.BACKEND_DWAVE_QPU: "dwave-qpu",
    miner_pb2.BACKEND_EXEC: "exec",
    miner_pb2.BACKEND_MOCK: "mock",
}
ALGORITHM_NAMES: dict[int, str] = {
    miner_pb2.ALGORITHM_UNSPECIFIED: "unspecified",
    miner_pb2.ALGORITHM_SA: "sa",
    miner_pb2.ALGORITHM_GIBBS: "gibbs",
    miner_pb2.ALGORITHM_QUANTUM_ANNEAL: "quantum-anneal",
    miner_pb2.ALGORITHM_FSA: "fsa",
    miner_pb2.ALGORITHM_MSA: "msa",
    miner_pb2.ALGORITHM_FLATIRON: "flatiron",
    miner_pb2.ALGORITHM_MPS: "mps",
    miner_pb2.ALGORITHM_MFA: "mfa",
    miner_pb2.ALGORITHM_SB: "sb",
    miner_pb2.ALGORITHM_BSB: "bsb",
    miner_pb2.ALGORITHM_GBSB: "gbsb",
    miner_pb2.ALGORITHM_GDSB: "gdsb",
    miner_pb2.ALGORITHM_GGDSB: "ggdsb",
    miner_pb2.ALGORITHM_HBSB: "hbsb",
    miner_pb2.ALGORITHM_HDSB: "hdsb",
    miner_pb2.ALGORITHM_SBQA: "sbqa",
    miner_pb2.ALGORITHM_TEDSB: "tedsb",
    miner_pb2.ALGORITHM_EXTERNAL: "external",
}

# Configure zeros mean "use the SDK default". These match
# quip_protocol::session::SessionConfig::from_configure.
DEFAULT_QUEUE_DEPTH = 3
DEFAULT_IDLE_TIMEOUT_S = 300
DEFAULT_HEARTBEAT_S = 15
DEFAULT_RECONNECT_WINDOW_S = 60
# Mirrors the Rust session loop's DEFAULT_NUM_SWEEPS in quip-solver-core.
DEFAULT_NUM_SWEEPS = 64


class MissingToken(Exception):
    pass


class BadWelcome(Exception):
    """Welcome.protocol_version is not what this SDK speaks."""

    def __init__(self, version: int):
        self.version = version
        super().__init__(f"unexpected protocol version in Welcome: {version}")


# Sysexits-style process exit codes (also carried on Fatal.exit_code).
EXIT_CLEAN = 0
EXIT_CONFIG_INVALID = 64
EXIT_ENV_INCOMPATIBLE = 69
EXIT_INTERNAL_FATAL = 70
EXIT_TOKEN_REJECTED = 77


@dataclass
class SessionConfig:
    miner_id: str
    queue_depth: int
    idle_timeout_s: int
    heartbeat_s: int
    reconnect_window_s: int
    # Session-wide sampling budget from Configure.backend_toml, applied when a
    # job does not pin its own num_sweeps. Mirrors the Rust session loop's
    # num_sweeps_from_toml (default 64).
    num_sweeps: int


def build_hello(miner_id: str, capabilities: miner_pb2.Capabilities) -> miner_pb2.Hello:
    """Build the miner Hello, reading QUIP_SESSION_TOKEN from the env."""
    token = os.environ.get("QUIP_SESSION_TOKEN")
    if not token:
        raise MissingToken("QUIP_SESSION_TOKEN unset")
    return miner_pb2.Hello(
        miner_id=miner_id,
        session_token=token,
        capabilities=capabilities,
    )


def check_welcome(welcome: miner_pb2.Welcome) -> None:
    """Reject a Welcome whose protocol_version is not what this SDK speaks."""
    version = welcome.protocol_version
    if version != PROTOCOL_VERSION:
        raise BadWelcome(version)


def num_sweeps_from_toml(backend_toml: str) -> int:
    """Read the session-wide `num_sweeps` from `Configure.backend_toml`.

    Matches the Rust session loop: only a top-level key counts (a `num_sweeps`
    under a backend's own `[table]` belongs to that backend), and anything
    absent or unusable falls back to DEFAULT_NUM_SWEEPS.
    """
    for line in backend_toml.splitlines():
        stripped = line.split("#", 1)[0].strip()
        if stripped.startswith("["):
            # First table header ends the top-level scope.
            break
        key, sep, value = stripped.partition("=")
        if sep and key.strip() == "num_sweeps":
            try:
                n = int(value.strip())
            except ValueError:
                break
            if n > 0:
                return n
            break
    return DEFAULT_NUM_SWEEPS


def session_config_from_configure(
    miner_id: str, configure: miner_pb2.Configure
) -> SessionConfig:
    def d(v: int, default: int) -> int:
        return default if v == 0 else v

    return SessionConfig(
        miner_id=miner_id,
        queue_depth=d(configure.queue_depth, DEFAULT_QUEUE_DEPTH),
        idle_timeout_s=d(configure.idle_timeout_s, DEFAULT_IDLE_TIMEOUT_S),
        heartbeat_s=d(configure.heartbeat_s, DEFAULT_HEARTBEAT_S),
        reconnect_window_s=d(configure.reconnect_window_s, DEFAULT_RECONNECT_WINDOW_S),
        num_sweeps=num_sweeps_from_toml(configure.backend_toml),
    )
